File: labs/control-ui/api-shadow/session_rotation_pool.ts
export type ManagedAccountStatus = 'READY' | 'IN_USE' | 'COOLING_DOWN' | 'LOCKED';

export interface ManagedAccountEntry {
  accountId: string;
  token: string;
  usesCount: number;
  maxUses: number;
  status: ManagedAccountStatus;
  cooldownUntilSec: number;
}

export class SessionRotationPool {
  private accounts = new Map<string, ManagedAccountEntry>();
  private rotationOrder: string[] = [];
  private cursor = 0;
  private defaultCooldownSec: number;
  constructor(defaultCooldownSec: number = 60) {
    this.defaultCooldownSec = defaultCooldownSec;
  }

  addAccount(accountId: string, token: string, maxUses: number): void {
    this.accounts.set(accountId, {
      accountId,
      token,
      usesCount: 0,
      maxUses,
      status: 'READY',
      cooldownUntilSec: 0,
    });
    this.rotationOrder.push(accountId);
  }

  acquireAccount(nowSec: number): { accountId: string; token: string } | null {
    if (this.rotationOrder.length === 0) return null;

    for (const acc of this.accounts.values()) {
      if (acc.status === 'COOLING_DOWN' && nowSec >= acc.cooldownUntilSec) {
        acc.status = 'READY';
        acc.usesCount = 0;
      }
    }

    const n = this.rotationOrder.length;
    for (let i = 0; i < n; i++) {
      const id = this.rotationOrder[this.cursor % n];
      this.cursor++;

      const acc = this.accounts.get(id!);
      if (acc && acc.status === 'READY') {
        acc.usesCount++;
        if (acc.usesCount >= acc.maxUses) {
          acc.status = 'COOLING_DOWN';
          acc.cooldownUntilSec = nowSec + this.defaultCooldownSec;
        }
        return { accountId: acc.accountId, token: acc.token };
      }
    }
    return null;
  }

  markLocked(accountId: string): void {
    const acc = this.accounts.get(accountId);
    if (acc) {
      acc.status = 'LOCKED';
    }
  }

  availableCount(nowSec: number): number {
    let count = 0;
    for (const acc of this.accounts.values()) {
      if (
        acc.status === 'READY' ||
        (acc.status === 'COOLING_DOWN' && nowSec >= acc.cooldownUntilSec)
      ) {
        count++;
      }
    }
    return count;
  }
}

File: labs/mobile/donor-archive/tunnel/SessionRotationPool.kt
package com.luminet.android.tunnel

enum class AndroidAccountStatus {
    READY, IN_USE, COOLING_DOWN, LOCKED
}

data class AndroidManagedAccount(
    val accountId: String,
    val token: String,
    var usesCount: Int = 0,
    val maxUses: Int = 10,
    var status: AndroidAccountStatus = AndroidAccountStatus.READY,
    var cooldownUntilSec: Long = 0
)

class SessionRotationPool(private val defaultCooldownSec: Long = 60) {
    private val accounts = mutableMapOf<String, AndroidManagedAccount>()
    private val rotationOrder = mutableListOf<String>()
    private var cursor = 0

    fun addAccount(accountId: String, token: String, maxUses: Int) {
        accounts[accountId] = AndroidManagedAccount(
            accountId = accountId,
            token = token,
            maxUses = maxUses
        )
        rotationOrder.add(accountId)
    }

    fun acquireAccount(nowSec: Long): Pair<String, String>? {
        if (rotationOrder.isEmpty()) return null

        for (acc in accounts.values) {
            if (acc.status == AndroidAccountStatus.COOLING_DOWN && nowSec >= acc.cooldownUntilSec) {
                acc.status = AndroidAccountStatus.READY
                acc.usesCount = 0
            }
        }

        val n = rotationOrder.size
        for (i in 0 until n) {
            val id = rotationOrder[cursor % n]
            cursor++

            val acc = accounts[id]
            if (acc != null && acc.status == AndroidAccountStatus.READY) {
                acc.usesCount++
                if (acc.usesCount >= acc.maxUses) {
                    acc.status = AndroidAccountStatus.COOLING_DOWN
                    acc.cooldownUntilSec = nowSec + defaultCooldownSec
                }
                return acc.accountId to acc.token
            }
        }
        return null
    }

    fun markLocked(accountId: String) {
        accounts[accountId]?.let { it.status = AndroidAccountStatus.LOCKED }
    }

    fun availableCount(nowSec: Long): Int {
        return accounts.values.count {
            it.status == AndroidAccountStatus.READY ||
                    (it.status == AndroidAccountStatus.COOLING_DOWN && nowSec >= it.cooldownUntilSec)
        }
    }
}
